Add beacon slot to EIP 1967 (#3)
OpenZeppelin/EIPs: EIPS/eip-1967.md

> ao  ---

Reviewer: Oops!

> Holds the address of the logic contract that this proxy delegates to.

Reviewer: ```suggestion
Holds the address of the logic contract that this proxy delegates to. SHOULD be empty if a beacon is used instead.
```

### Beacon contract address

Storage slot `0xa3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50` (obtained as `bytes32(uint256(keccak256('eip1967.proxy.beacon')) - 1)`).

> Holds the address of the beacon contract this proxy relies on. This is an alternative to the logic contract slot. See the **Beacons** section below.

Reviewer: I think we have to add something like "optional" here, and/or to specify that if both beacon and implementation are present which one should take precedence.

Author: This is what I tried to mean when I say its an alternative to the logic slot. I'll try to rephrase that/

> Holds the address of the beacon contract this proxy relies on (fallback). This is an alternative to the logic contract slot, and should only be considered if the logic contract slot is not set. See the **Beacons** section below.

Reviewer: ```suggestion
Holds the address of the beacon contract this proxy relies on (fallback). SHOULD be empty if a logic address is used directly instead, and should only be considered if the logic contract slot is empty. See the Beacons section below.
```

From [_Malicious backdoors in Ethereum proxies_](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357) by Nomic Labs:

> Any function in the Proxy contract whose selector matches with one in the implementation contract will be called directly, completely skipping the implementation code.
>
> Because the function selectors use a fixed amount of bytes, there will always be the possibility of a clash. This isn’t an issue for day to day development, given that the Solidity compiler will detect a selector clash within a contract, but this becomes exploitable when selectors are used for cross-contract interaction. Clashes can be abused to create a seemingly well-behaved contract that’s actually concealing a backdoor.

The fact that proxy public functions are potentially exploitable makes it necessary to standardise the logic contract address in a different way. This approach is also used as part of [Universal Upgradeable Proxy Standard (EIP1822)](./eip-1822.md), which could become a specialization of this EIP.

## Beacons

> Some use-cases rely on multiple proxy contracts delegating their calls to the same logic contract. If each contract was to store the address of the logic contract using the logic contract storage slot, upgrading them would require writing to the corresponding storage slot of every instance, which would be very inefficient.

Reviewer 2: ```suggestion
Some use-cases rely on multiple proxy contracts delegating their calls to the same logic contract. If each contract was to store the address of the logic contract using the logic contract storage slot, upgrading them would require writing to the corresponding storage slot of every instance, which could be very expensive in terms of gas.
```

> Another approach is to have each proxy retrieve the logic contract's address for a dedicated beacon. Using this pattern, the address of the logic contract can be modified in the beacon and take immediate effect for all the corresponding proxy contracts.

Reviewer: Beacon in quotes because it's the first time the term is introduced.

```suggestion
Another approach is to have each proxy retrieve the logic contract's address from a dedicated "beacon". Using this pattern, the address of the logic contract can be modified in the beacon and take immediate effect for all the corresponding proxy contracts.
```

> A beacon MUST implements the function:

Reviewer: ```suggestion
A beacon MUST implement the function:
```

```
function implementation() returns (address)
```

Beacon based proxy contracts do not use the logic contract slot. Instead, they use the beacon contract slot to store the address of the beacon they are attached to. In order to know the logic contract used by a beacon proxy, one should:

- Read the address of the beacon for the beacon logic storage slot;

> - Call the `implementation()` function on the beacon contract (with `msg.sender` set to be the proxy's address).

Reviewer: Have you seen `msg.sender` used in a beacon?

I don't think this should be part of the specification. Off-chain you can simulate the sender using `eth_call`, but on-chain other contracts would not be able to retrieve the implementation if `msg.sender` changed the result. If anything I would specify that `implementation()` "MUST NOT" depend on `msg.sender`, or  "SHOULD NOT" if we want to allow people to do it if they weigh the consequences.
